### PR TITLE
Improve `admin.vm.volume.Import` service

### PR DESCRIPTION
**Pull request by DemiMarie**
The main improvement is that a qube cannot cause dom0 to use a huge amount of memory, but there are other improvements as well.  See individual commit messages for details.

### CODE REVIEW

**Comment by marmarek** [comment]
```
DEBUG: Sphinx error:
DEBUG: Undocumented arguments for command 'qubesd-query': '--max-bytes, --single-line'
```

**Comment by marmarek** [comment]
Any specific reason for adding `isdigit()` call before every `int()` on a string? The latter will reject non-numeric strings (also if there is just non-numeric prefix-suffix) with `ValueError` exception, which I think is handled in all the places you modified.

File: qubes/tools/qubesd_query.py [hunk 1]
```diff
@@ -75,20 +87,56 @@ async def qubesd_client(socket, payload, *args):
...
 
+# pylint: disable=too-many-return-statements
 def main(args=None):
     args = parser.parse_args(args)
     loop = asyncio.get_event_loop()
+    max_payload_size = 1024 if args.single_line else MAX_PAYLOAD_SIZE
+
+    if args.max_bytes is not None:
+        if args.max_bytes > max_payload_size:
+            parser.error('Maximum payload too large (max {})'
```
Comment (marmarek) [inline qubes/tools/qubesd_query.py:98]: 64k is fine for the default limit, but may be a bit low for upper bound what is accepted at all. Anyway, preventing the user to raise the limit sounds pointless.

**Comment by DemiMarie** [comment]
> Any specific reason for adding `isdigit()` call before every `int()` on a string?

The call to `isdigit()` should be on `bytes`, not on `str`, as the latter is Unicode-aware (which is not wanted).

> The latter will reject non-numeric strings (also if there is just non-numeric prefix-suffix) with `ValueError` exception, which I think is handled in all the places you modified.

`int()` accepts arbitrary Python base-10 integer literals, including unwanted ones such as literals with leading or trailing whitespace.

File: qubes/tools/qubesd_query.py (hunk 1 above)
Comment (marmarek) [reply qubes/tools/qubesd_query.py:98]: Ah, right.

**Comment by marmarek** [comment]
> `int()` accepts arbitrary Python base-10 integer literals, including unwanted ones such as literals with leading or trailing whitespace.

I don't think that's any practical problem here. Anyway, if you really insist, it sounds like this wants `validate_and_parse_int()` utility function.

File: qubes/api/__init__.py [hunk 2]
```diff
@@ -210,6 +210,17 @@ def enforce(predicate):
         if not predicate:
             raise PermissionDenied()
 
+    @classmethod
```
Comment (marmarek) [inline qubes/api/__init__.py:213]: You call it on an object instance, not class type, so it should be normal method.

File: qubes/api/__init__.py [hunk 3]
```diff
@@ -210,6 +210,17 @@ def enforce(predicate):
...
+    def validate_size(cls, untrusted_size: bytes) -> int:
+        cls.enforce(isinstance(untrusted_size, bytes))
+        if not untrusted_size.isdigit():
+            raise qubes.api.ProtocolError('Size must be ASCII digits (only)')
+        if len(untrusted_size) >= 20:
+            raise qubes.api.ProtocolError('Sizes limited to 19 decimal digits')
```
Comment (marmarek) [inline qubes/api/__init__.py:219]: this case is not covered by tests

File: qubes/api/__init__.py [hunk 4]
```diff
@@ -210,6 +210,17 @@ def enforce(predicate):
...
+        try:
+            return int(untrusted_size)
+        except ValueError as e:
+            raise qubes.api.ProtocolError(e.args[0])
```
Comment (marmarek) [inline qubes/api/__init__.py:223]: and this too

File: qubes/tools/qubesd_query.py [hunk 5]
```diff
@@ -75,20 +87,56 @@ async def qubesd_client(socket, payload, *args):
...
+                if 0x20 <= untrusted_c <= 0x7E:
+                    untrusted_v.append(untrusted_c)
+                else:
+                    parser.error('Invalid byte {} in input'.format(untrusted_c))
+                    # make sure to terminate, even if parser.error() would
+                    # return for some reason
+                    return 1
+            else:
+                parser.error('No newline found in first {} bytes'
+                             .format(args.max_bytes))
```
Comment (marmarek) [inline qubes/tools/qubesd_query.py:125]:
```suggestion
                             .format(max_payload_size))
```
`self.max_bytes` is `None` usually...

**Review by marmarek — CHANGES_REQUESTED** [review]
Typo in the last commit message

File: qubes/tools/qubesd_query.py [hunk 6]
```diff
@@ -35,6 +36,20 @@
...
+parser.add_argument('--single-line',
+    dest='single_line',
+    action='store_true',
+    default=False,
+    help='Only read a single line of input')
+parser.add_argument('--null',
+    dest='null',
+    action='store_true',
+    default=False,
+    help="Print an extra NUL byte; useful for shell scripts")
```
Comment (marmarek) [inline qubes/tools/qubesd_query.py:52]: Can you clarify what's the point of this? Not only the option, but also printing this byte at all. Preferably in man page's description of this option (which is missing, for other options added here too).

**Review by marmarek — CHANGES_REQUESTED** [review]
Besides the comment below, unit tests fail.

File: qubes/tools/qubesd_query.py (hunk 6 above)
Comment (DemiMarie) [reply qubes/tools/qubesd_query.py:52]: I added documentation for this option.  The other new options have documentation already IIUC.

File: qubes/tests/__init__.py [hunk 7]
```diff
@@ -383,6 +383,39 @@ def __exit__(self, exc_type, exc_value, tb):
...
+        if exc_info is None:
+            continue
+        ex = exc_info[1]
+        while ex is not None:
+            if isinstance(ex, qubes.exc.QubesVMError):
+                ex.vm = None
+            traceback.clear_frames(ex.__traceback__)
+            ex = ex.__context__
+
+class QubesTestResult(unittest.TestResult):
```
Comment (marmarek) [inline qubes/tests/__init__.py:397]: I appreciate your attempt at cleaning this up, but a) it does not belong to this PR, and b) this most likely will fail with alternative runners as they may provide own result class (we do use `nose2` and sometimes `pytest` example).

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
```
************* Module qubes.tools.qubesd_query
qubes/tools/qubesd_query.py:164:10: E0001: Parsing failed: 'expected ':' (<unknown>, line 164)' (syntax-error)
```

File: qubes/tools/qubesd_query.py [hunk 8]
```diff
@@ -75,20 +92,56 @@ async def qubesd_client(socket, payload, *args):
...
+                # make sure to terminate, even if parser.error() would return
+                # for some reason
+                return 1
+            payload = bytes(untrusted_v)
+            del untrusted_c, untrusted_v
+        else:
+            # read one byte more to check for too long payload,
+            # instead of silently truncating
+            payload = sys.stdin.buffer.read(MAX_PAYLOAD_SIZE + 1)
+            if len(payload) > args.max_bytes:
```
Comment (marmarek) [inline qubes/tools/qubesd_query.py:140]: Tests say:
```
Traceback (most recent call last):
  File "/usr/bin/qubesd-query", line 5, in <module>
    sys.exit(main())
  File "/usr/lib/python3.8/site-packages/qubes/tools/qubesd_query.py", line 140, in main
    if len(payload) > args.max_bytes:
TypeError: '>' not supported between instances of 'int' and 'NoneType'
```

**Review by marmarek — CHANGES_REQUESTED** [review]

**Comment by marmarek** [comment]
> @DemiMarie
[Avoid relying on undocumented unittest internals](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/e725af8687017d2b6ff5df3658096a6851d4d287)
[e725af8](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/e725af8687017d2b6ff5df3658096a6851d4d287)
@DemiMarie
[Wrap coroutine in Task before passing to asyncio.wait](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/aafbf6c11f10ae2cea7b30ec7ee13accc7eb35c6)
[aafbf6c](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/aafbf6c11f10ae2cea7b30ec7ee13accc7eb35c6)

I requested you to not mix those with other functional changes...

**Comment by DemiMarie** [comment]
> > @DemiMarie
> > [Avoid relying on undocumented unittest internals](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/e725af8687017d2b6ff5df3658096a6851d4d287)
> > [e725af8](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/e725af8687017d2b6ff5df3658096a6851d4d287)
> > @DemiMarie
> > [Wrap coroutine in Task before passing to asyncio.wait](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/aafbf6c11f10ae2cea7b30ec7ee13accc7eb35c6)
> > [aafbf6c](https://github.com/QubesOS/qubes-core-admin/pull/500/commits/aafbf6c11f10ae2cea7b30ec7ee13accc7eb35c6)
> 
> I requested you to not mix those with other functional changes...

oops, fixed

(My local branch includes these commits so that the tests work, but I should not have pushed them.)

**Comment by DemiMarie** [comment]
PipelineRetryFailed

**Review by marmarek — APPROVED** [review]
